src/editors/containers/ProblemEditor/data/OLXParser.js

```diff
@@ -270,37 +271,61 @@ export class OLXParser {
+      const hasNonQuestionKeys = (obj) => !(_.some(nonQuestionKeys, (key) => _.has(obj, key)));
+      const isWhiteSpace = (obj) => _.get(obj, '#text', 'nonemptystring').trim() !== '';
+      const transformIntoQuestionObject = (root) => {
+        return _.filter(_.filter(root, isWhiteSpace), hasNonQuestionKeys);
+      };
+      let questionObject = {};
+
+      // try to find question within problemType node; odd path because of `preserveOrder` option used for parsing
+      const problemTypeObject = _.get(
+        _.find(parsedOlx[0].problem, (obj, index) => _.has(obj, problemType)),
+        problemType,
+        {},
+      );
+      questionObject = transformIntoQuestionObject(problemTypeObject);
+
+      // if haven't found any question nodes, check if question is stored in the problem node
+      if (_.isEmpty(questionObject)) {
+        questionObject = transformIntoQuestionObject(parsedOlx[0].problem);
+      }
+      return questionObject;
     };
 
-    /* Only numerical response has different ways to generate OLX, test with
-      numericInputWithFeedbackAndHintsOLXException and numericInputWithFeedbackAndHintsOLX
-      shows the different ways the olx can be generated.
-    */
-    if (_.isArray(problemObject)) {
-      questionObject = _.omitBy(problemObject[0], (value, key) => _.includes(nonQuestionKeys, key));
-    } else {
-      questionObject = _.omitBy(problemObject, (value, key) => _.includes(nonQuestionKeys, key));
-    }
-    // Check if problem tag itself will have question and descriptions.
-    if (_.isEmpty(questionObject)) {
-      questionObject = _.omitBy(this.problem, (value, key) => _.includes(nonQuestionKeys, key));
-    }
-    const serializedQuestion = _.mapKeys(questionObject, (value, key) => _.get(tagMap, key, key));
+    const mapQuestionTags = (questionObject) => {
+      /* TODO: How do we uniquely identify the label and description?
+        In order to parse label and description, there should be two states
+        and settings should be introduced to edit the label and description.
+        In turn editing the settings update the state and then it can be added to
+        the parsed OLX.
+      */
+      const tagMap = {
+        label: 'bold',
+        description: 'em',
+      };
 
-    const questionString = builder.build(serializedQuestion);
-    return questionString;
+      return _.map(questionObject, (obj) => _.mapKeys(obj, (value, key) => _.get(tagMap, key, key)));
+    };
+
+    const parser = new XMLParser({
+      preserveOrder: true,
+      trimValues: false,
+      ignoreAttributes: false,
+    });
+    const parsedOlx = parser.parse(this.olxString);
+
+    const questionObject = getQuestionObject(parsedOlx);
+    const serializedQuestion = mapQuestionTags(questionObject);
+
+    const builder = new XMLBuilder({
+      preserveOrder: true,
+      attributeNamePrefix: '@_',
+      ignoreAttributes: false,
+      format: false,
+    });
+    return builder.build(serializedQuestion);
```

Comment:
We should add comments to explain why are we doing this.

Reply:
@farhaanbukhsh Can you help me understand what would be the best comment here? Nothing comes to my mind other than "that is what is necessary for getting the correct HTML output from the OLX", probably because I've written the code myself.

What things do you think seem confusing without additional context to why we are doing that?

Reply:
Comment: 

```
Parsing the olxstring separately with preserveOrder: true so that we can find out the question in the OLX. Later we build the OLX by preserving styling in the questions.
```

Let me know @Cup0fCoffee if this makes sense to you :)

Reply:
Thank you! I've added the comment that might be a bit verbose, but explains well why this is necessary.